### PR TITLE
Added search filter using substring match strategy

### LINKED ISSUE #63: [Feature Request] Firebase does not support in-depth filtering. Please provide a solution for filtering data efficiently.
Example: Social Media Posts

I want to filter posts by IDs (reported post IDs). I have an array of IDs that I want to use to filter posts from a Firebase collection on the client side.

### PR DESCRIPTION
**Pull request by delwar36**
Just add a filter option to filter by substring of any filed on the document.

```
String _searchText = 'match me';
FirestorePagination(
.....
    searchFilter: FilterModel(
        fieldName: 'title',
        searchValue: _searchText,
    ),
.....
)
```

### CODE REVIEW

**Comment by OutdatedGuy** [comment]
Hi @delwar36, can you explain the need for this PR? I believe this can be achieved by just adding a filter in the `itemBuilder` itself.

e.g.

```dart
FirestorePagination(
  itemBuilder: (context, snapshot, index) {
    final title = snapshot[index].get('title');

    if (!title.contains('match me')) {
      return SizedBox.shrink();
    }

    // return your widget here
  },
),
```

**Comment by delwar36** [comment]
> Hi @delwar36, can you explain the need for this PR? I believe this can be achieved by just adding a filter in the `itemBuilder` itself.
> 
> e.g.
> 
> ```dart
> FirestorePagination(
>   itemBuilder: (context, snapshot, index) {
>     final title = snapshot[index].get('title');
> 
>     if (!title.contains('match me')) {
>       return SizedBox.shrink();
>     }
> 
>     // return your widget here
>   },
> ),
> ```

It works on `ViewType.list` type. But in `ViewType.grid` type it leaves blank spaces as grid dilegates dynamically creates items sizes. We maybe solve this problem using custom deligates though.

**Comment by OutdatedGuy** [comment]
@delwar36 got it. I'll review the changes when I get some free time.

File: example/pubspec.lock
Comment (OutdatedGuy) [inline example/pubspec.lock:1]: Revert changes to this file. I'll update the example is another PR

File: lib/firebase_pagination.dart [hunk 1]
```diff
@@ -12,3 +12,4 @@ export 'src/firestore_pagination.dart';
 // Data Models
 export 'src/models/view_type.dart';
 export 'src/models/wrap_options.dart';
+export 'src/models/filter_model.dart';
```
Comment (OutdatedGuy) [inline lib/firebase_pagination.dart]: Move this import above `export 'src/models/view_type.dart';` to follow alphabetical order

File: lib/src/models/filter_model.dart
Comment (OutdatedGuy) [inline lib/src/models/filter_model.dart:1]: Add dart doc comments (`///`) for each of the public members to get intellisense description of the usage.

Refer other classes to match the documentation style.

File: pubspec.yaml
Comment (OutdatedGuy) [inline pubspec.yaml:1]: Why are changes made here? Please revert this.

**Review by OutdatedGuy — CHANGES_REQUESTED** [review]
Thanks for the PR. I have requested some changes to get this PR approved.

File: example/pubspec.lock [hunk 2]
```diff
@@ -310,4 +310,4 @@ packages:
     version: "0.5.1"
 sdks:
   dart: ">=3.4.3 <4.0.0"
-  flutter: ">=3.18.0-18.0.pre.54"
+  flutter: ">=3.18.0-18.0.pre.54"
```
Comment (OutdatedGuy) [inline example/pubspec.lock]: I guess you accidentally removed the last empty line here

File: pubspec.yaml [hunk 3]
```diff
@@ -52,4 +52,4 @@ flutter:
   #         weight: 700
   #
   # For details regarding fonts in packages, see
-  # https://flutter.dev/custom-fonts/#from-packages
+  # https://flutter.dev/custom-fonts/#from-packages
```
Comment (OutdatedGuy) [inline pubspec.yaml]: I guess you accidentally removed the last empty line here

File: lib/src/models/filter_model.dart [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class FilterModel {
```
Comment (OutdatedGuy) [inline lib/src/models/filter_model.dart:2]: Also add documentation for the class here

File: lib/src/models/filter_model.dart [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+  /// The name of the field in collection to be searched.
+  final String fieldName;
+
+  /// The substring that is intented to search in [fieldName] value.
+  final String searchValue;
+
+  /// Creates a object that contains the properties of the [FilterModel] model.
```
Comment (OutdatedGuy) [inline lib/src/models/filter_model.dart]: Change this to `Creates a [FilterModel] object.` and move the constructor declaration to the top of class.

**Review by OutdatedGuy — CHANGES_REQUESTED** [review]
Just few more nitpicks and it's good to go

File: example/pubspec.lock [hunk 6]
```diff
@@ -311,3 +311,4 @@ packages:
 sdks:
   dart: ">=3.4.3 <4.0.0"
   flutter: ">=3.18.0-18.0.pre.54"
+
```
Comment (OutdatedGuy) [inline example/pubspec.lock]: Something seems to be messed up here.

Screenshot 2024-11-11 at 8 45 25 PM

File: pubspec.yaml [hunk 7]
```diff
@@ -53,3 +53,4 @@ flutter:
   #
   # For details regarding fonts in packages, see
   # https://flutter.dev/custom-fonts/#from-packages
+
```
Comment (OutdatedGuy) [inline pubspec.yaml]: Something seems to be messed up here.

Screenshot 2024-11-11 at 8 45 43 PM

**Review by OutdatedGuy — CHANGES_REQUESTED** [review]

File: example/pubspec.lock (hunk 6 above)
Comment (delwar36) [reply example/pubspec.lock]: There was an indentation on the last blank line

File: pubspec.yaml (hunk 7 above)
Comment (delwar36) [reply pubspec.yaml]: There was an indentation on the last blank line

**Review by OutdatedGuy — APPROVED** [review]
LGTM!

**Pull request merged**